### PR TITLE
[docker-outside-of-docker] - Fix externally-managed-environment python error for "bookworm" - solution to issue #1120

### PR DESCRIPTION
**Pull request by gauravsaini04**
**Feature Name**
* Docker-outside-of-docker

**Description**
* This PR tries to raise a solution for features issue #1120 

**_Changelog_**
* Added required code in `install.sh` file and added a test case scenario for testing it

**Checklist**
* [x] Checked that applied changes work as expected

### CODE REVIEW

File: test/docker-outside-of-docker/scenarios.json [hunk 1]
```diff
@@ -142,5 +142,15 @@
                 "mobyBuildxVersion": "0.12.0"
             }
         }
+    },
+    "docker_python_bookworm": {
+        "image": "mcr.microsoft.com/devcontainers/base:bookworm",
```
Comment (samruddhikhandale) [inline test/docker-outside-of-docker/scenarios.json:147]: Thanks! 🎉

File: src/docker-outside-of-docker/install.sh [hunk 2]
```diff
@@ -302,35 +307,38 @@ fi
...
-            find_version_from_git_tags compose_v1_version "https://github.com/docker/compose" "tags/"
-            echo "(*) Installing docker-compose ${compose_v1_version}..."
-            curl -fsSL "https://github.com/docker/compose/releases/download/${compose_v1_version}/docker-compose-Linux-x86_64" -o /usr/local/bin/docker-compose
-            chmod +x /usr/local/bin/docker-compose
+            echo "(*) Installing docker compose v1 via pip..."
+            export PYTHONUSERBASE=/usr/local
+            pip3 install --disable-pip-version-check --no-cache-dir --user "Cython<3.0" pyyaml wheel docker-compose --no-build-isolation
         fi
     else
         echo "(*) Installing compose-switch as docker-compose..."
```
Comment (samruddhikhandale) [inline src/docker-outside-of-docker/install.sh:344]: Can you also update the `else` condition to match https://github.com/devcontainers/features/blob/main/src/docker-in-docker/install.sh#L368-L390?  Thanks!

Comment (samruddhikhandale) [reply src/docker-outside-of-docker/install.sh:344]: Also, add a new option `INSTALL_DOCKER_COMPOSE_SWITCH`  by referring tohttps://github.com/devcontainers/features/commit/4db22d38c6f04996de99ff7062555e0eb2f66b9c#diff-a72a4e49c44e54f09feb84b7767d355d68d34fbee561d8f96a99d12d2488b9c5R19

File: src/docker-outside-of-docker/devcontainer-feature.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
     "id": "docker-outside-of-docker",
-    "version": "1.5.0",
+    "version": "1.5.1",
```
Comment (samruddhikhandale) [inline src/docker-outside-of-docker/devcontainer-feature.json:3]:
```suggestion
    "version": "1.6.0",
```

Given the new additions

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/docker-outside-of-docker/install.sh (hunk 2 above)
Comment (gauravsaini04) [reply src/docker-outside-of-docker/install.sh]: corrected. thanks for inputs !

File: src/docker-outside-of-docker/install.sh [hunk 4]
```diff
@@ -17,6 +17,7 @@ SOURCE_SOCKET="${SOURCE_SOCKET:-"/var/run/docker-host.sock"}"
 TARGET_SOCKET="${TARGET_SOCKET:-"/var/run/docker.sock"}"
 USERNAME="${USERNAME:-"${_REMOTE_USER:-"automatic"}"}"
 INSTALL_DOCKER_BUILDX="${INSTALLDOCKERBUILDX:-"true"}"
+INSTALL_DOCKER_COMPOSE_SWITCH="${INSTALLDOCKERCOMPOSESWITCH:-"true"}"
```
Comment (samruddhikhandale) [inline src/docker-outside-of-docker/install.sh:20]: Shouldn't you make changes to `src/docker-outside-of-docker/devcontainer-feature.json` and publicize this Feature option?

Comment (samruddhikhandale) [reply src/docker-outside-of-docker/install.sh:20]: Use https://github.com/devcontainers/features/blob/main/src/docker-in-docker/devcontainer-feature.json#L54-L58 for reference

Comment (samruddhikhandale) [inline src/docker-outside-of-docker/install.sh:20]: Can we add tests to validate this new `installDockerComposeSwitch` option?

File: src/docker-outside-of-docker/install.sh [hunk 5]
```diff
@@ -273,6 +279,22 @@ if [ "${USE_MOBY}" = "true" ]; then
     fi
 fi
 
+
+docker_home="/usr/libexec/docker"
+cli_plugins_dir="${docker_home}/cli-plugins"
+
+fallback_compose(){
```
Comment (samruddhikhandale) [inline src/docker-outside-of-docker/install.sh:286]:
```suggestion
install_compose_fallback(){
```

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/docker-outside-of-docker/install.sh (hunk 4 above)
Comment (gauravsaini04) [reply src/docker-outside-of-docker/install.sh:20]: yes forgot, made changes

Comment (gauravsaini04) [reply src/docker-outside-of-docker/install.sh:20]: added tests for validation of installation and when not installed

File: src/docker-outside-of-docker/install.sh (hunk 5 above)
Comment (gauravsaini04) [reply src/docker-outside-of-docker/install.sh]: made changes

File: src/docker-outside-of-docker/install.sh [hunk 6]
```diff
@@ -273,6 +279,22 @@ if [ "${USE_MOBY}" = "true" ]; then
...
 
+
+docker_home="/usr/libexec/docker"
+cli_plugins_dir="${docker_home}/cli-plugins"
+
+install_compose_fallback(){
+    local url=$1
+    local compose_version=$2
+    local target_compose_arch=$3
+    local docker_compose_path=$4
```
Comment (samruddhikhandale) [inline src/docker-outside-of-docker/install.sh:290]: Do we need to explicitly pass these args?

File: src/docker-outside-of-docker/install.sh [hunk 7]
```diff
@@ -302,47 +324,86 @@ fi
...
+        # Setup v1 CLI as alternative in addition to compose-switch (which maps to v2)
+        mv "${current_compose_path}" "${target_compose_path}"
+        update-alternatives --install ${docker_compose_path} docker-compose /usr/local/bin/compose-switch 99
+        update-alternatives --install ${docker_compose_path} docker-compose "${target_compose_path}" 1
+    else
+        err "Skipping installation of compose-switch as docker compose is unavailable..."
     fi
 fi
 
+
```
Comment (samruddhikhandale) [inline src/docker-outside-of-docker/install.sh:406]:
```suggestion
```

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/docker-outside-of-docker/install.sh (hunk 7 above)
Comment (gauravsaini04) [reply src/docker-outside-of-docker/install.sh]: done

File: src/docker-outside-of-docker/install.sh (hunk 6 above)
Comment (gauravsaini04) [reply src/docker-outside-of-docker/install.sh]: ya not reqd. , done

**Review by samruddhikhandale — APPROVED** [review]
Wonderful, thank you!

**Pull request merged**